File: app/src/main/java/com/example/sw0b_001/Models/GatewayClients/GatewayClient.java
package com.example.sw0b_001.Models.GatewayClients;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.PrimaryKey;

@Entity(indices = {@Index(value={"MSISDN"}, unique = true)})
public class GatewayClient {
    @PrimaryKey(autoGenerate = true)
    public long id;

    @ColumnInfo(name="type")
    String type;

    @ColumnInfo(name="MSISDN")
    String MSISDN;

    @ColumnInfo(name="default")
    boolean isDefault = false;

    @ColumnInfo(name="operator_name")
    String operatorName;

    @ColumnInfo(name="last_ping_session")
    double lastPingSession = 0.0;

    public GatewayClient(String type, String MSISDN, String operatorName, boolean isDefault) {
        this.type = type;
        this.MSISDN = MSISDN;
        this.operatorName = operatorName;
        this.isDefault = isDefault;
    }

    public GatewayClient() {}

    public long getId() {
        return id;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getMSISDN() {
        return MSISDN;
    }

    public void setMSISDN(String MSISDN) {
        this.MSISDN = MSISDN;
    }

    public String getOperatorName() {
        return this.operatorName;
    }

    public void setOperatorName(String operatorName) {
        this.operatorName = operatorName;
    }

    public boolean isDefault() {
        return this.isDefault;
    }

    public void setDefault(boolean isDefault) {
        this.isDefault = isDefault;
    }

    public void setLastPingSession(double lastPingSession) {
        this.lastPingSession = lastPingSession;
    }

    public double getLastPingSession() {
        return this.lastPingSession;
    }
}

File: app/src/main/java/com/example/sw0b_001/Models/GatewayClients/GatewayClient.kt
package com.example.sw0b_001.Models.GatewayClients

import android.content.Context
import android.widget.Toast
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import com.example.sw0b_001.Database.Datastore
import com.example.sw0b_001.Models.Publisher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.net.URL

@Entity(indices = [Index(value = ["MSISDN"], unique = true)])
class GatewayClient {
    @PrimaryKey(autoGenerate = true)
    var id: Long = 0

    @ColumnInfo(name = "type")
    var type: String? = null

    @ColumnInfo(name = "MSISDN")
    var mSISDN: String? = null

    @ColumnInfo(name = "default")
    var isDefault: Boolean = false

    @ColumnInfo(name = "operator_name")
    var operatorName: String? = null

    var alias: String? = null

    @ColumnInfo(name = "operator_id")
    var operatorId: String? = null

    @ColumnInfo(name = "country")
    var country: String? = null

    @ColumnInfo(name = "last_ping_session")
    var lastPingSession: Double = 0.0

    @ColumnInfo(defaultValue = "0")
    var date: Long = System.currentTimeMillis()

    constructor(
        type: String?,
        MSISDN: String?,
        operatorName: String?,
        country: String?,
        isDefault: Boolean
    ) {
        this.type = type
        this.mSISDN = MSISDN
        this.operatorName = operatorName
        this.isDefault = isDefault
        this.country = country
    }

    constructor()

    companion object {
        var TYPE_CUSTOM: String = "custom"

        fun refreshGatewayClients(context: Context, failedCallback: Runnable) {
            val scope = CoroutineScope(Dispatchers.Default)
            scope.launch {
                try {
                    Publisher.getAvailablePlatforms(context, failedCallback).let{ json ->
                        json.forEach { it->
                            val url = URL(it.icon_png)
                            it.logo = url.readBytes()
                        }
                        Datastore.getDatastore(context).availablePlatformsDao()
                            .fetchAllList().forEach {
                                if(!json.contains(it)) {
                                    Datastore.getDatastore(context).availablePlatformsDao()
                                        .delete(it.name)
                                }
                            }
                        Datastore.getDatastore(context).availablePlatformsDao()
                            .insertAll(json)
                    }
                } catch(e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }
}
